### ISSUE TITLE
[BUG] Optuna-Integration no longer depends on pytorch-lightning

### ISSUE THREAD

**Issue by daidahao** [unit8co/darts#3093]
**Describe the bug**
`optuna-integration[pytorch-lightning]` now depends on `lightning` and not `pytorch-lightning` any more. Running `uv sync --group dev-all --upgrade` will result in an error due to `lightning` not being available on win32 py312 (the package does exist on win32 py312, but there might be markers from other packages that prevent the resolution.)

**To Reproduce**
Steps to reproduce the behavior, preferably code snippet.

```sh
uv sync --group dev-all --upgrade
  × No solution found when resolving dependencies for split (markers: python_full_version == '3.12.*' and platform_machine != 's390x' and sys_platform ==
  │ 'win32'):
  ╰─▶ Because there are no versions of lightning and optuna-integration[pytorch-lightning]>=4.7.0 depends on lightning, we can conclude that
      optuna-integration[pytorch-lightning]>=4.7.0 cannot be used.
      And because only the following versions of optuna-integration[pytorch-lightning] are available:
          optuna-integration[pytorch-lightning]<=4.7.0
          optuna-integration[pytorch-lightning]==4.8.0
      we can conclude that optuna-integration[pytorch-lightning]>=4.7.0,<4.8.0 cannot be used. (1)

      Because there are no versions of lightning and optuna-integration[pytorch-lightning]==4.8.0 depends on lightning, we can conclude that
      optuna-integration[pytorch-lightning]==4.8.0 cannot be used.
      And because we know from (1) that optuna-integration[pytorch-lightning]>=4.7.0,<4.8.0 cannot be used, we can conclude that
      optuna-integration[pytorch-lightning]>=4.7.0 cannot be used.
      And because darts:dev-all depends on optuna-integration[pytorch-lightning]>=4.7.0 and your project requires darts:dev-all, we can conclude that your
      project's requirements are unsatisfiable.

      hint: While the active Python version is 3.13, the resolution failed for other Python versions supported by your project. Consider limiting your
      project's supported Python versions using `requires-python`.
```

Immediate suggested solution is to change `optional` dependency group from:
```
optuna-integration[pytorch-lightning]>=4.7.0
```
to
```
optuna-integration>=4.7.0
```

However, a longer-term solution might be to address #2198, i.e., migrate from `pytorch-lightning` to `lightning`.

**Expected behavior**

**System (please complete the following information):**
 - Python version: 3.13.12
 - darts version: latest source
 - macOS: 26.3
 - uv: 0.10.2

**Additional context**

**Comment by daidahao** [comment]
Update: `lightning` on PyPI was compromised on 2026-04-30. The new versions `2.6.2` and `2.6.3` were pulled soon after, hence the resolution error here. I've created a PR #3096 to insert a `uv` cooldown period and mitigate this kind of risk in the future.

Anyone that has installed  `2.6.2` or `2.6.3` on 2026-04-30 should rotate all their credentials. I've checked my environment and did not find trace of either version, but will rotate credentials just to be on the safe side.

Update: The compromised `lightning` should only affect developers who might have run `uv sync --group dev-all --upgrade` during the brief hour `lightning` was breached, and does not affect users because `lightning` is not part of optional dependencies whatsoever. 

I have also checked with the maintainer of Darts and can confirm that Darts is not affected.

**Comment by dennisbader** [comment]
We discussed offline: We decided against migrating to lightning and instead implement the `PyTorchLightningPrunerCallback` from `optuna-integration[pytorch-lightning]` directly in Darts and drop that dependency entirely.

**Comment by eastani** [comment]
Hi @dennisbader, I'd like to take this on along the direction you outlined.

Sketching the plan so it's easy to push back on before I open a PR:

- **Vendor the callback into Darts** as `darts.utils.callbacks.PyTorchLightningPruningCallback`, importing from `pytorch_lightning` directly (which sidesteps the existing `PatchedPruningCallback` workaround in `tests/optional_deps/test_optuna.py`). Optuna is treated as an optional dependency — the class is importable without it, and instantiation raises a clear error if optuna is missing.
- **Update existing callers**: `tests/optional_deps/test_optuna.py` drops the multiple-inheritance workaround and uses the Darts callback directly; `examples/17-hyperparameter-optimization.ipynb` is updated to import from `darts.utils.callbacks`.
- **Add unit tests** covering at minimum: importability without optuna, instantiation, warning when monitor metric is missing, and pruning behaviour on a tiny Torch model (mirroring the existing optuna integration test).
- **Drop `optuna-integration[pytorch-lightning]>=4.7.0`** from `pyproject.toml` optional deps.
- **CHANGELOG** entry under Unreleased → For users → Dependencies.

Will open a draft PR for review once the implementation is green locally. Happy to adjust direction if any of the above isn't what you had in mind.
